contrib/format-esri/src/main/java/org/apache/drill/exec/store/esri/ShpBatchReader.java

```diff
@@ -0,0 +1,334 @@
+      Matcher matcher = pattern.matcher(wktReference);
+      if (matcher.find()) {
+        this.srid = Integer.parseInt(matcher.group(1));
+        this.spatialReference = SpatialReference.create(srid);
+      }
+
+      logger.debug("Processing Shape File : {}", hadoopShp);
+    } catch (IOException e) {
+      throw UserException
+        .dataReadError(e)
+        .message("Failed to open open input file: %s", split.getPath())
+        .addContext("User name", negotiator.userName())
+        .build(logger);
+    }
+  }
+
+  private String byteArrayToString(byte[] in) {
+    char out[] = new char[in.length * 2];
+    for (int i = 0; i < in.length; i++) {
+      out[i * 2] = "0123456789ABCDEF".charAt((in[i] >> 4) & 15);
+      out[i * 2 + 1] = "0123456789ABCDEF".charAt(in[i] & 15);
+    }
+    return new String(out);
+  }
+
+  private void processShapefileSet(RowSetLoader rowWriter, final int gid, final Geometry geom, final Object[] dbfRow) {
+    rowWriter.start();
+
+    gidWriter.setInt(gid);
+    sridWriter.setInt(srid);
+    shapeTypeWriter.setString(geom.getType().toString());
+    final ByteBuffer buf = (ByteBuffer) OGCGeometry.createFromEsriGeometry(geom, spatialReference).asBinary();
+    final byte[] bytes = buf.array();
+    geomWriter.setBytes(bytes, bytes.length);
+    writeDbfRow(dbfRow, rowWriter);
+
+    rowWriter.save();
+  }
+
+  private void writeDbfRow(final Object[] dbfRow, RowSetLoader rowWriter) {
+    int dbfFieldCount = dbfReader.getHeader().getFieldsCount();
+
+    for (int i = 0; i < dbfFieldCount; i++) {
+      DbfField field = dbfReader.getHeader().getField(i);
+
+      if (dbfRow[i] == null) {
+        continue;
+      }
+
+      switch (field.getDataType()) {
+        case CHAR:
+          byte[] strType = (byte[]) dbfRow[i];
+          String stringValue = new String( strType, Charset.forName("utf-8")).trim();
+          writeStringColumn(rowWriter, field.getName(), stringValue);
+          break;
+        case FLOAT:
+          final double floatVal = ((Float) dbfRow[i]).doubleValue();
+          writeDoubleColumn(rowWriter, field.getName(), floatVal);
+          break;
+        case DATE:
+          final long dataVal = ((java.util.Date) dbfRow[i]).getTime();
+          writeTimeColumn(rowWriter, field.getName(), dataVal);
+          break;
+        case LOGICAL:
+          int logicalVal = (Boolean) dbfRow[i] ? 1 : 0;
+          writeBooleanColumn(rowWriter, field.getName(), logicalVal);
+          break;
+        case NUMERIC:
+          double numericVal = ((Number) dbfRow[i]).doubleValue();
+          if (field.getDecimalCount() == 0) {
+            writeIntColumn(rowWriter, field.getName(), (int) numericVal);
+          } else {
+            writeDoubleColumn(rowWriter, field.getName(), numericVal);
+          }
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  private void writeStringColumn(TupleWriter rowWriter, String name, String value) {
```

Comment:
Isn't the set of columns fixed? If so, do we need to do the hash table lookup each time?

Reply:
See comment above.  There are 4 or 5 known fields and the rest are optional.  By optional I mean that they can be literally anything.